### ISSUE TITLE
Question: Putting Assets in a Different Folder

### ISSUE THREAD

**Issue by jaredsilver** [gatsbyjs/gatsby#8833]
## Summary

I'm a growth engineer @DataCamp, and I'm running an experiment that involves rewriting our home page in Gatsby to measure gains in performance and conversion metrics.

We plan to reverse proxy the homepage on datacamp.com to split 50% of traffic between our current application and our new Gatsby application.

The problem we're running into is that every JS, CSS, and other file that Gatsby generates would also need to be reverse proxied in order for this to work.

So far as I've been able to see, Gatsby does not provide a mechanism to store such files in a separate directory (which would make it easy to reverse proxy all of the traffic there).

Does Gatsby (or a plugin) provide a way to store all of these files in a single directory?

Alternatively, does Gatsby provide a way to prefix calls to these assets if they are being hosted elsewhere (i.e. on a separate subdomain)?

## Relevant information

Potentially relevant past issues, PRs, and plugins: #6346, https://github.com/madecomfy/gatsby-plugin-asset-path

**Comment by mikelax** [comment]
One possible solution would be to use the [path prefix feature](https://www.gatsbyjs.org/docs/path-prefix/#path-prefix) of gatsby. If you go this route, one thing to consider is that any `<Link>` component you use will be relative to the gatsby site (ie will respect the path prefix when deployed. So if you are linking to other pages you will need to keep that in mind. 

The other option is to deploy the gatsby site to a totally separate separate web server, so then your reverse proxy just needs to route to one server or the other.

**Comment by jaredsilver** [comment]
Thanks, @gatsbot. This will hopefully be addressed in #10933, but leaving it open until then 🤞

**Comment by jaredsilver** [comment]
Also, for anyone who comes across this and is wondering how we solved it in the meantime: we used `onCreateWebpackConfig` in `gatsby-node.js` to manually prepend a string to all filenames and chunkfilenames, thus allowing us to reverse proxy on a regular expression including that string. However, this is awful, and I'm hoping #10933 will eliminate the need for us to continue doing it this way 😄

**Comment by siddv** [comment]
Heya @jaredsilver . I'm trying to do something similar. Do you have an example of how you did this?

**Comment by jaredsilver** [comment]
@siddv Yup! Here's part of the code we're currently using in `gatsby-node.js`. It's pretty bad and makes a lot of assumptions. I'm hoping not to have to refactor it though and am waiting to see what happens with #10933 😂

The string that we use to identify a file that should be routed to AWS rather than our server is `marketingasset`. It's just hardcoded in there for now.

**Note: I'm assuming this is very brittle and has the potential to break in surprising ways with little to no warning.**

```js
const path = require('path');

exports.onCreateWebpackConfig = ({
  stage,
  rules,
  loaders,
  plugins,
  actions,
  getConfig,
  setWebpackConfig,
}) => {
  if (stage === 'build-javascript') {
    let config = getConfig();
    config.output.filename = 'marketingasset-[name]-[contenthash].js';
    config.output.chunkFilename = 'marketingasset-[name]-[contenthash].js';

    function replaceFilePathsInOptions(options) {
      if (options.name) {
        options.name = 'static/marketingasset-[name]-[hash].[ext]';
      } else if (options.localIdentName) {
        options.localIdentName =
          'marketingasset--[name]--[local]--[hash:base64:5]';
      }
    }

    config.module.rules.map(rule => {
      if (rule.oneOf) {
        rule.oneOf.map(one => {
          if (one.use) {
            one.use.map(use => {
              if (use.options) {
                replaceFilePathsInOptions(use.options);
              }
              return use;
            });
          }
          return one;
        });
      } else if (rule.use) {
        rule.use.map(use => {
          if (use.options) {
            replaceFilePathsInOptions(use.options);
          }
          return use;
        });
      }
      return rule;
    });

    config.plugins.map(plugin => {
      if (plugin.options) {
        if (plugin.options.filename) {
          plugin.options.filename = 'marketingasset-[name].[contenthash].css"';
        }
        if (plugin.options.chunkFilename) {
          plugin.options.chunkFilename =
            'marketingasset-[name].[contenthash].css';
        }
      }
    });

    actions.replaceWebpackConfig(config);
  }
};
```

**Comment by siddv** [comment]
Thanks a lot @jaredsilver! We were trying something similar, but using paths and trying to move the asset files into subfolders. Was causing _a lot_ of issues. Renaming the files seems to work _much_ better. Thanks for this!

**Comment by Jonic** [comment]
@jaredsilver This looks great - we've been struggling with this recently too, so we're going to give this a try! Thanks :)

**Comment by jaredsilver** [comment]
ℹ️Functionality will be added in https://github.com/gatsbyjs/gatsby/pull/12128. Thanks, @DSchau!

**Comment by jaredsilver** [comment]
I don’t seem to have permissions to reopen the issue 😢

@DSchau could you reopen it?

Also, is Gatsbot malfunctioning? This issue definitely had activity within the past 30 days!

**Comment by fodkastir** [comment]
@jaredsilver Hi Jared, we are having a similar issue and I am trying to put all assets to a different sub-folder like public/assets. I am wondering if this is what you were trying to achieve? Or you wanted to put them into a different folder like /assets? The reason I asked is that it seems to me that my goal could be achieved by doing some custom webpack config but I am trying to confirm that before I start to dig into it. Thanks in advance!

**Comment by DSchau** [comment]
Published and available in gatsby@2.4.0. Specifically, the following packages have been augmented with assetPrefix functionality.

- gatsby@2.4.0
- gatsby-link@2.1.0 (gatsby@2.4.0 depends upon this, don't need to install manually!)
- gatsby-plugin-feed@2.2.0
- gatsby-plugin-manifest@2.1.0
- gatsby-plugin-offline@2.1.0
- gatsby-plugin-sitemap@2.1.0

Soon enough, the documentation will be available at [/docs/asset-prefix/](https://gatsbyjs.org/docs/asset-prefix/) (waiting for the build to complete), but you can get a sneak peek [here](https://github.com/gatsbyjs/gatsby/tree/master/docs/docs)
